### ISSUE TITLE
Gravity forms merge tags error in form notifications

### ISSUE THREAD

**Issue by Twansparant** [roots/acorn#198]
### Terms

- [X] I have read the [guidelines for Contributing to Roots Projects](https://github.com/roots/.github/blob/master/CONTRIBUTING.md)
- [X] This request is not a duplicate of an existing issue
- [X] I have read the [docs](https://roots.io/docs/) and followed them (if applicable)
- [X] I have seached the [Roots Discourse](https://discourse.roots.io/) for answers and followed them (if applicable)
- [X] This is not a personal support request that should be posted on the [Roots Discourse](https://discourse.roots.io/) community

#### What's wrong?
I can't seem to insert any Gravity forms merge tags on the forms notification edit screens. It it working on the forms confirmation edit screens.

#### What have you tried?
- I disabled all other plugins except Gravity forms itself
- I switched form my Sage theme to the default Twenty Twenty-Two theme
- Disabled the `\Roots\bootloader()` call in `functions.php`
- Disabled all of my wordpress actions & filters
- Disabled all of my enqueued javascripts
- Moved the composer dependency from my theme to Bedrock 

#### What insights have you gained?
- Switching to default theme fixes the issue, but that's only because it's not loading Acorn with the bootloader.  

#### Possible solutions
- Remove Acorn as dependency
 
#### Temporary workarounds
- None

### Steps To Reproduce

1. Install the Gravity forms plugin
2. Create a form and go to it's notifications page
3. Try to insert a merge tag in the from email address field

### Expected Behavior

I expect a dropdown list with the all form fields merge tags to appear.

### Actual Behavior

Nothing happens, I get an error in my browser console.

### Relevant Log Output

```shell
Uncaught TypeError: Cannot read properties of null (reading 'value')
    at MaybeAddSaveLinkMergeTag (admin.php:1227:54)
    at admin.php:78:22
    at Array.forEach (<anonymous>)
    at Object.doHook (admin.php:70:11)
    at Object.applyFilters (admin.php:38:17)
    at gfMergeTagsObj.getMergeTags (form_admin.js:1322:21)
    at gfMergeTagsObj.getMergeTagListItems (form_admin.js:1463:24)
    at HTMLAnchorElement.<anonymous> (form_admin.js:952:35)
    at HTMLAnchorElement.dispatch (jquery.js:5430:27)
    at HTMLAnchorElement.elemData.handle (jquery.js:5234:28)
```

### Versions

2.0.0

**Comment by Twansparant** [comment]
I know it doesn't look like one, but I can re-produce this with a bare installation of:

- Trellis
- Bedrock
- Acorn
- Sage 10

With no modifications whatsoever.

So I boiled it down to the usage of **Acorn**. When I switch from the **Sage** starter theme to the **Twenty Twenty-Two** theme, the error is gone.

**Comment by Twansparant** [comment]
True, but when I disable the `Register The Bootloader` & `Register Sage Theme Files` calls in **Sage**, the error is also gone?

`\Roots\bootloader();`

**Comment by edwinsiebel** [comment]
I'm not using roots, so not familiar with that.

However, if Laravel is loaded (with the helpers file), the error occured at my end.

Maybe the register bootloader loads this specific file?

**Comment by QWp6t** [comment]
Thanks for the explanation @edwinsiebel.

Yes, Acorn loads Laravel helpers.

@Twansparant this is a bug in Gravity Forms.

A workaround on your end could be to create a global `event()` function that loads before Acorn loads; the function would be in charge of handling calls from GF and Acorn. Out of the box, Acorn doesn't use that function, so you might be able to get away with assuming that only GF will call the function.

Feel free to make a post about this on Roots Discourse: https://discourse.roots.io/

**Comment by Twansparant** [comment]
@edwinsiebel Thanks for your reply!
That makes sense. I also reported this to gravity forms, but since it was not occuring on the default theme, I thought later that it must be caused on my end and not theirs.

**Comment by Twansparant** [comment]
@QWp6t Check, will try that and I'll report back!
There must be quit a few people though using Laravel & Gravityforms?
Thanks for the tips!

**Comment by Twansparant** [comment]
> A workaround on your end could be to create a global `event()` function that loads before Acorn loads; the function would be in charge of handling calls from GF and Acorn.

The problem is, that the function shouldn't be callable in the first place, so even if I override the `event` function, it will still fail in returning the value.

**Comment by Twansparant** [comment]
So this is not related to acorn indeed, since it also happens with any of the default themes by just adding a function called _event_ to to **functions.php**:

```php
function event() {
	return true;
}
```

**Comment by Twansparant** [comment]
Nope, no fix unfortunately...
I filed a bug report to GravityForms but they don't think it's a bug on their end, so I could only add a `product suggestion`

**Comment by Twansparant** [comment]
I'm still having this issue, and I still don't know how to fix it...
I tried something like this in my **functions.php** that basically adds an `event` function that shouldn't only be callable once by `Acorn` and then never again:

```php
global $event_called;
$event_called = false;

if (!$event_called) {
	function event() {
		global $event_called;
		$event_called = true;
	}
}

try {
	\Roots\bootloader();
} catch (Throwable $e) {
	wp_die(
		__('You need to install Acorn to use this theme.', 'sage'),
		'',
		[
			'link_url' => 'https://docs.roots.io/acorn/2.x/installation/',
			'link_text' => __('Acorn Docs: Installation', 'sage'),
		]
	);
}
```
And in my **app/Providers/ThemeServiceProvider.php**:

```php
class ThemeServiceProvider extends ServiceProvider {
	public function boot() {
		event();
	}
}
```

When I log `function_exists('event')` in [Acorn's helpers file](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Foundation/helpers.php#L439) it returns `true`, so that works.

When I log `is_callable('event')` in the GravityForms `get_attributes` function in `includes/settings/fields/class-base.php` on **L328** it also returns `true`, so that doesn't work.

So basically how can you make a function _uncallable_, is that even possible?

**Comment by tombroucke** [comment]
I ran across the same issue today. I found an easy workaround, by just redefining the `MaybeAddSaveLinkMergeTag` function with a better dom query (which doesn't rely on the id attribute) later on the page.

```php
add_action('admin_footer', function () {
    $isGravityFormsEditPage = isset($_GET['page']) && 'gf_edit_forms' === $_GET['page'];
    if (!$isGravityFormsEditPage) {
        return;
    }
    ?>
    <script type="text/javascript">
    function MaybeAddSaveLinkMergeTag( mergeTags, elementId, hideAllFields, excludeFieldTypes, isPrepop, option ) {
        const eventSelectEl = document.querySelector('select[name="_gform_setting_event"]');
        if(!eventSelectEl) {
            return mergeTags;
        }
        var event = eventSelectEl.value;
        if ( event === 'form_saved' || event === 'form_save_email_requested' ) {
            mergeTags['other'].tags.push( {
                tag:  '{save_link}',
                label: <?php echo json_encode(esc_html__('Save & Continue Link', 'gravityforms')); ?>
            } );
            mergeTags['other'].tags.push( {
                tag:   '{save_token}',
                label: <?php echo json_encode(esc_html__('Save & Continue Token', 'gravityforms')); ?>
            } );
        }
        return mergeTags;
    }
    </script>
    <?php
});
```

**Comment by Twansparant** [comment]
Thanks @tombroucke, gonna try that first thing in the morning!

**Comment by Twansparant** [comment]
Yes your fix works like a charm @tombroucke
Many thanks!

**Comment by christianmagill** [comment]
@tombroucke , Is this still the best solution we have?
